### PR TITLE
[FLINK-27163] Add label for the session job to help list the session …

### PR DESCRIPTION
**Pull request by Aitozi**
…jobs in the same session cluster

This PR is meant to add label for the session job to help to list session jobs under the same session.

image

### CODE REVIEW

**Comment by morhidi** [comment]
LGTM overall. nit: Shouldn't we use a shorter label, e.g: 
```
k get sessionjob -l target=basic-session-cluster
```

**Comment by morhidi** [comment]
On second thought we could use something like this for filtering, aren't we?
```
k get --field-selector spec.deploymentName=basic-session-cluster
```

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/sessionjob/FlinkSessionJobReconciler.java [hunk 1]
```diff
@@ -76,6 +78,24 @@ public void reconcile(FlinkSessionJob flinkSessionJob, Context context) throws E
                 OperatorUtils.getSecondaryResource(
                         flinkSessionJob, context, configManager.getOperatorConfiguration());
 
+        var labels = flinkSessionJob.getMetadata().getLabels();
```
Comment (morhidi) [inline flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/sessionjob/FlinkSessionJobReconciler.java:81]: I suggest to extract this into a utility method (we could use computeIfAbsent)

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/sessionjob/FlinkSessionJobReconciler.java [hunk 2]
```diff
@@ -76,6 +78,24 @@ public void reconcile(FlinkSessionJob flinkSessionJob, Context context) throws E
...
+        if (labels == null) {
+            labels = new HashMap<>();
+        }
+        if (!flinkSessionJob
+                .getSpec()
+                .getDeploymentName()
+                .equals(labels.get(LABEL_SESSION_CLUSTER))) {
+            labels.put(LABEL_SESSION_CLUSTER, flinkSessionJob.getSpec().getDeploymentName());
+            flinkSessionJob.getMetadata().setLabels(labels);
+            kubernetesClient
```
Comment (morhidi) [inline flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/sessionjob/FlinkSessionJobReconciler.java:91]: Can't we use an update control for this?

**Comment by Aitozi** [comment]
> k get sessionjob --field-selector spec.deploymentName=basic-session-cluster

I tried this way before, but the custom resource seems do not support field selector

```
kubectl get sessionjob --field-selector spec.deploymentName=basic-session-cluster
Error from server (BadRequest): Unable to find "flink.apache.org/v1beta1, Resource=flinksessionjobs" that match label selector "", field selector "spec.deploymentName=basic-session-cluster": field label not supported: spec.deploymentName

kubectl version --short
Flag --short has been deprecated, and will be removed in the future. The --short output will become the default.
Client Version: v1.24.0
Kustomize Version: v4.5.4
Server Version: v1.23.3
```

After search the related issues, I think the custom resource can only support field selector by `metadata.namespace` and `metadata.name`

https://github.com/kubernetes/kubernetes/pull/53345

**Comment by Aitozi** [comment]
> LGTM overall. nit: Shouldn't we use a shorter label, e.g:
> 
> ```
> k get sessionjob -l target=basic-session-cluster
> ```
> 
> It's easier to remember :)

Good idea, but the `target` is a bit too generic, I lean to rename it to `target.session`

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/sessionjob/FlinkSessionJobReconciler.java (hunk 2 above)
Comment (Aitozi) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/sessionjob/FlinkSessionJobReconciler.java]: The reconciler do not return `UpdateControl` now, If we want to use the update control for it, we should adjust the interface first to return the update control :/ . We have just decide not to do [this](https://issues.apache.org/jira/browse/FLINK-26915) now. I'm not sure whether we need to break it for this PR.

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/sessionjob/FlinkSessionJobReconciler.java (hunk 1 above)
Comment (Aitozi) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/sessionjob/FlinkSessionJobReconciler.java]: Moved to a dedicated method.

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/sessionjob/FlinkSessionJobReconciler.java (hunk 2 above)
Comment (morhidi) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/sessionjob/FlinkSessionJobReconciler.java]: Got it, but this is the first use case that we modify something else then the status. We might need to reconsider this again. Updating the CR deep in the stack sounds worse to me.

**Comment by morhidi** [comment]
> > k get sessionjob --field-selector spec.deploymentName=basic-session-cluster
> 
> I tried this way before, but the custom resource seems do not support field selector
> 
> ```
> kubectl get sessionjob --field-selector spec.deploymentName=basic-session-cluster
> Error from server (BadRequest): Unable to find "flink.apache.org/v1beta1, Resource=flinksessionjobs" that match label selector "", field selector "spec.deploymentName=basic-session-cluster": field label not supported: spec.deploymentName
> 
> kubectl version --short
> Flag --short has been deprecated, and will be removed in the future. The --short output will become the default.
> Client Version: v1.24.0
> Kustomize Version: v4.5.4
> Server Version: v1.23.3
> ```
> 
> After search the related issues, I think the custom resource can only support field selector by `metadata.namespace` and `metadata.name`
> 
> [kubernetes/kubernetes#53345](https://github.com/kubernetes/kubernetes/pull/53345)


> > k get sessionjob --field-selector spec.deploymentName=basic-session-cluster
> 
> I tried this way before, but the custom resource seems do not support field selector
> 
> ```
> kubectl get sessionjob --field-selector spec.deploymentName=basic-session-cluster
> Error from server (BadRequest): Unable to find "flink.apache.org/v1beta1, Resource=flinksessionjobs" that match label selector "", field selector "spec.deploymentName=basic-session-cluster": field label not supported: spec.deploymentName
> 
> kubectl version --short
> Flag --short has been deprecated, and will be removed in the future. The --short output will become the default.
> Client Version: v1.24.0
> Kustomize Version: v4.5.4
> Server Version: v1.23.3
> ```
> 
> After search the related issues, I think the custom resource can only support field selector by `metadata.namespace` and `metadata.name`
> 
> [kubernetes/kubernetes#53345](https://github.com/kubernetes/kubernetes/pull/53345)

Yeah, I was almost sure you've tried it, but I thought it was worth asking :)

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/sessionjob/FlinkSessionJobReconciler.java (hunk 2 above)
Comment (Aitozi) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/sessionjob/FlinkSessionJobReconciler.java]: Yes, agree with it. I'm ok to rework the interface to work out it. Also cc @gyfora @wangyang0918

Comment (morhidi) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/sessionjob/FlinkSessionJobReconciler.java]: We hit some race conditions anyway with the status updates, so it maybe good as is with the java-operator-sdk v2 and we can change it later.

**Comment by morhidi** [comment]
+1
For the record we need this functionality due: [Enable arbitrary CRD field selectors](https://github.com/kubernetes/kubernetes/issues/53459)

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/sessionjob/FlinkSessionJobReconciler.java (hunk 2 above)
Comment (Aitozi) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/sessionjob/FlinkSessionJobReconciler.java]: As I know we have manually patch the status after https://github.com/apache/flink-kubernetes-operator/pull/199 . So this change do not bring the resourceVersion conflicts, I think. Do you mean keep it as now ?

Comment (morhidi) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/sessionjob/FlinkSessionJobReconciler.java]: Let's see what @gyfora and @wangyang0918 think/suggest.

**Comment by wangyang0918** [comment]
TBH, I do not like to do the CR `.metadata` or `.spec` updates in the reconciler. Especially, this is only for easier filter, not for operator internal use case. In the production environment, I believe the users could do this manually or covered by external job management tools.

I also have another feeling this is the responsibility of [mutating admission webhook](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#mutatingadmissionwebhook).

**Comment by morhidi** [comment]
> TBH, I do not like to do the CR `.metadata` or `.spec` updates in the reconciler. Especially, this is only for easier filter, not for operator internal use case. In the production environment, I believe the users could do this manually or covered by external job management tools.
> 
> I also have another feeling this is the responsibility of [mutating admission webhook](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#mutatingadmissionwebhook).

I've been also thinking about putting this into the webhook or maybe using the label exclusively for linking the two CRs.

**Comment by Aitozi** [comment]
@wangyang0918 Regarding why we need this: although we can manage the job with external job management tools, but we also may need to operate in command line. In the case to operate on all the jobs of a session cluster, eg: delete all the session jobs before shutdown the cluster, it will be useful I think.

I think the mutating admission webhook is a better way. But one shortcoming is it's optional. If disabled, it will not work

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/sessionjob/FlinkSessionJobReconciler.java (hunk 2 above)
Comment (gyfora) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/sessionjob/FlinkSessionJobReconciler.java]: if you don't explicitly set the resourceversion to `null` before patching you might still get an error.

**Comment by gyfora** [comment]
I am also leaning toward -1 on the approach in this PR as so far we have explicitly avoided making changes to the metadata/spec for other features.

**Comment by gyfora** [comment]
I think mutating webhooks provide a good alternative solution but I would prefer to revisit this issue after the release.

**Comment by Aitozi** [comment]
Thanks for your good suggestion @morhidi @wangyang0918 @gyfora . I will try to implement it with mutating webhook. It's ok to me to leave it for next release

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/sessionjob/FlinkSessionJobReconciler.java (hunk 2 above)
Comment (Aitozi) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/sessionjob/FlinkSessionJobReconciler.java]: yeah, but I think this operation is retriable

**Comment by gyfora** [comment]
Should we close this PR for now?

**Pull request closed without merging**